provider/src/test/kotlin/org/gradle/kotlin/dsl/integration/ScriptModelIntegrationTest.kt

```diff
@@ -72,10 +72,74 @@ abstract class ScriptModelIntegrationTest : AbstractIntegrationTest() {
         assertThat(sourcePathFor(subProjectScriptFile).map { it.name }, matches)
     }
 
-    private
+    protected
     fun sourcePathFor(scriptFile: File) =
         kotlinBuildScriptModelFor(projectRoot, scriptFile).sourcePath
 
+    protected
+    class ProjectSourceRoots(val projectDir: File, val sourceSets: List<String>, val languages: List<String>)
+
+    protected
+    fun withMainSourceSetJavaIn(projectDir: String) =
+        ProjectSourceRoots(existing(projectDir), listOf("main"), listOf("java"))
+
+    protected
+    fun withMainSourceSetJavaKotlinIn(projectDir: String) =
+        ProjectSourceRoots(existing(projectDir), listOf("main"), listOf("java", "kotlin"))
+
+    protected
+    fun matchesProjectsSourceRoots(vararg projectSourceRoots: ProjectSourceRoots): Matcher<Iterable<File>> {
+
+        fun hasLanguageDir(base: File, set: String, lang: String): Matcher<Iterable<*>> =
+            hasItem(base.resolve("src/$set/$lang"))
+
+        val all = mutableListOf<Matcher<Iterable<*>>>()
```

Comment:
:x: It might not be terribly efficient but I would go with the collection pipeline version:

```kotlin

    protected
    fun matchesProjectsSourceRoots(vararg projectSourceRoots: ProjectSourceRoots): Matcher<Iterable<File>> {

        fun hasLanguageDir(base: File, set: String, lang: String): Matcher<Iterable<*>> =
            hasItem(base.resolve("src/$set/$lang"))

        return allOf(
            *projectSourceRoots
                .filter { it.languages.isNotEmpty() }
                .flatMap { sourceRoots ->
                    val languageDirs =
                        sourceRoots.sourceSets.flatMap { sourceSet ->
                            listOf("java", "kotlin").map { language ->
                                val hasLanguageDir = hasLanguageDir(sourceRoots.projectDir, sourceSet, language)
                                if (language in sourceRoots.languages) hasLanguageDir
                                else not(hasLanguageDir)
                            }
                        }

                    val resourceDirs =
                        sourceRoots.sourceSets.map { sourceSet ->
                            hasLanguageDir(sourceRoots.projectDir, sourceSet, "resources")
                        }

                    languageDirs + resourceDirs
                }.toTypedArray())
    }
```